flow/multiagent_envs/traffic_light_grid.py

```diff
@@ -0,0 +1,260 @@
+                   ),
+            dtype=np.float32)
+        return tl_box
+
+    @property
+    def action_space(self):
+        """See class definition."""
+        if self.discrete:
+            return Discrete(2)
+        else:
+            return Box(
+                low=-1,
+                high=1,
+                shape=(1,),
+                dtype=np.float32)
+
+    def get_state(self):
+        """Observations for each traffic light agent.
+
+        :return: dictionary which contains agent-wise observations as follows:
+        - For the self.num_observed number of vehicles closest and incoming
+        towards traffic light agent, gives the vehicle velocity, distance to
+        intersection, edge_number, density traffic light state.
+        - For edges in the network, gives the density and average velocity.
+        - For the self.num_local_lights number of nearest lights (itself
+        included), gives the traffic light information, including the last
+        change time, light direction (i.e. phase), and a currently_yellow flag.
+        """
+        # Normalization factors
+        max_speed = max(
+            self.k.scenario.speed_limit(edge)
+            for edge in self.k.scenario.get_edge_list())
+        grid_array = self.net_params.additional_params["grid_array"]
+        max_dist = max(grid_array["short_length"], grid_array["long_length"],
+                       grid_array["inner_length"])
+
+        # TODO(cathywu) refactor PO_TrafficLightGridEnv with convenience
+        # methods for observations, but remember to flatten for single-agent
+
+        # Observed vehicle information
+        speeds = []
+        dist_to_intersec = []
+        edge_number = []
+        all_observed_ids = []
+        for _, edges in self.scenario.node_mapping:
+            local_speeds = []
+            local_dists_to_intersec = []
+            local_edge_numbers = []
+            for edge in edges:
+                observed_ids = \
+                    self.k_closest_to_intersection(edge, self.num_observed)
+                all_observed_ids.append(observed_ids)
+
+                # check which edges we have so we can always pad in the right
+                # positions
+                local_speeds.extend(
+                    [self.k.vehicle.get_speed(veh_id) / max_speed for veh_id in
+                     observed_ids])
+                local_dists_to_intersec.extend([(self.k.scenario.edge_length(
+                    self.k.vehicle.get_edge(
+                        veh_id)) - self.k.vehicle.get_position(
+                    veh_id)) / max_dist for veh_id in observed_ids])
+                local_edge_numbers.extend([self._convert_edge(
+                    self.k.vehicle.get_edge(veh_id)) / (
+                    self.k.scenario.network.num_edges - 1) for veh_id in
+                                           observed_ids])
+
+                if len(observed_ids) < self.num_observed:
```

Comment:
So, it seems to me that there's also an edge with number zero. So, appending a tuple of zeros to the end could also correspond to a vehicle on edge zero with zero speed and zero distance to intersection. Thoughts?

Reply:
Yeah, good call. I'm changing the defaults to 1, which implies vehicles which are fast and far away.